aptos-move/aptos-vm-types/src/change_set.rs

```diff
@@ -151,69 +258,108 @@ impl VMChangeSet {
                 },
                 Vacant(entry) => {
                     // This is a new write op. It can overwrite a delta so we
                     // have to make sure we remove such a delta from the set in
                     // this case.
-                    let removed_delta = delta_change_set.remove(entry.key());
+                    let removed_delta = aggregator_delta_set.remove(entry.key());
 
                     // We cannot create after modification with a delta!
-                    if removed_delta.is_some() && next_write_op.is_creation() {
+                    if removed_delta.is_some() && additional_write_op.is_creation() {
                         return Err(VMStatus::error(
                             StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
                             err_msg("Cannot create a resource after modification with a delta."),
                         ));
                     }
 
-                    entry.insert(next_write_op);
+                    entry.insert(additional_write_op);
                 },
             }
         }
 
-        let write_set = write_set_mut.freeze().map_err(|_| {
-            VMStatus::error(
-                StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
-                err_msg("Error when freezing squashed write sets."),
-            )
-        })?;
+        Ok(())
+    }
 
-        // Squash events.
-        events.extend(next_events);
+    fn squash_additional_writes(
+        write_set: &mut BTreeMap<StateKey, WriteOp>,
+        additional_write_set: BTreeMap<StateKey, WriteOp>,
+    ) -> anyhow::Result<(), VMStatus> {
+        for (key, additional_write_op) in additional_write_set.into_iter() {
+            match write_set.entry(key) {
+                Occupied(mut entry) => {
+                    squash_writes_pair!(entry, additional_write_op);
+                },
+                Vacant(entry) => {
+                    entry.insert(additional_write_op);
+                },
+            }
+        }
+        Ok(())
+    }
 
-        Self::new(write_set, delta_change_set, events, checker)
+    pub fn squash_additional_change_set(
+        self,
+        additional_change_set: Self,
+        checker: &dyn CheckChangeSet,
+    ) -> anyhow::Result<Self, VMStatus> {
+        // First, obtain write sets, delta change sets and events of this and
+        // additional change sets.
+        let Self {
+            resource_write_set: additional_resource_write_set,
+            module_write_set: additional_module_write_set,
+            aggregator_write_set: additional_aggregator_write_set,
+            aggregator_delta_set: additional_aggregator_delta_set,
+            events: additional_events,
+        } = additional_change_set;
+        let Self {
+            mut resource_write_set,
+            mut module_write_set,
+            mut aggregator_write_set,
+            mut aggregator_delta_set,
+            mut events,
+        } = self;
```

Comment:
why do we need to unpack here? because we can't pass two &mut in the squash function?